### PR TITLE
feat: add initial wasmcloud wasi kv implementation

### PR DESCRIPTION
**Pull request by vados-cosmonic**
## Feature or Problem

Add an implementation [wasi keyvalue contract](https://github.com/WebAssembly/wasi-keyvalue) that allows building fused components.

See the rendered Markdown README for a more thorough explanation.

While the code in there right now errors for all keyvalue interface bits, I will be changing that to `wasmbus`

## Related Issues

N/A

## Release Information

`componentize` v1.0.0

## Consumer Impact

Consumers can try out the component model with wasmCloud

## Testing

N/A

Built on platform(s)
- [x] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [ ] aarch64-darwin
- [ ] x86_64-windows

Tested on platform(s)
- [x] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [ ] aarch64-darwin
- [ ] x86_64-windows

### Unit Test(s)

N/A

### Acceptance or Integration

N/A

### Manual Verification

N/A

### CODE REVIEW

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+
+    pub static INPUT_STREAM: RefCell<InputStream> = RefCell::new(0);
+    pub static INPUT_STREAMS: RwLock<HashMap<InputStream, IncomingValue>> = RwLock::new(HashMap::new());
+
+    pub static OUTPUT_STREAM: RefCell<OutputStream> = RefCell::new(0);
+    pub static OUTPUT_STREAMS: RwLock<HashMap<OutputStream, OutgoingValue>> = RwLock::new(HashMap::new());
+
+    pub static BUCKET: RefCell<Bucket> = RefCell::new(0);
+    pub static BUCKETS: RwLock<HashMap<Bucket, String>> = RwLock::new(HashMap::new());
+}
```
Comment (brooksmtownsend) [inline wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:42]: Could you comment a little here on the use of input/output streams to handle these keyvalue operations? It's a bit of a different pattern than I'm used to (just directly handling the operation)

File: wasmcloud-wasi-kv/README.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
+# `wasmcloud-wasi-kv`
+
+This repository contains a [WebAssembly System Interface (WASI) key-value contract][wasi-kv] compliant Web Assembly module which can be composed into a "macro" or "composite" component to provide key value functionality.
```
Comment (brooksmtownsend) [inline wasmcloud-wasi-kv/README.md:3]: Are macro/composite components the official language?

Also nit: `Web Assembly` -> `WebAssembly`

File: wasmcloud-wasi-kv/README.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+This module is special because it implements the KV contract for any and all [wasmCloud][wasmcloud] environments, showcasing the reusability of WASI, WIT ,  it's .
+
+## ❓ What does that even mean?
+
+- [WebAssembly ("WASM")][wasm] brings amazing sandboxing and almost-C/C++/Rust speeds
+- Compiling your favorite `$LANGUAGE` to WebAssembly modules means you can run anywhere WASM modules run
+- WASI introduces a framework for extending the functionality available to any given WebAssembly sandbox
+- [Wasm Interface Type ("WIT")][wit] is the language (an [IDL][idl], like gRPC/Thrift/etc) for describing functionality
+- The [WASI key-value contract][wasi-kv] describes operations a cache key value store (think [Redis][redis]) would offer
+- If a WebAssembly module implements the key-value contract, it can be *composed*/*combined* with another module that *depends on* that contract.
```
Comment (brooksmtownsend) [inline wasmcloud-wasi-kv/README.md:14]: I'm going to suggest some rewordings here because I'm having a hard time following, and I think there's a lot of work to do for someone first seeing these notions. I'd love to see these bullets be something like (and please feel free to wordsmith)

- WebAssembly / Wasm is a compilation target for many languages that creates a platform agnostic binary that can run on any CPU architecture, operating system, or even in the browser
- Webassembly System interface (WASI) is a framework that extends platform-specific APIs to the WebAssembly sandbox, allowing Wasm to do things like operate on a filesystem or make use of network sockets
- [Wasm Interface Type ("WIT")][wit] is an interface definition language for describing functionality that WebAssembly modules can use
- The WASI key-value contract describes basic operations (get, set, del) for interacting with a key-value store. This can be implemented in many different ways, this example uses Redis

TL;DR: WebAssembly components allows you to use a common interface to write an application that uses a key-value store: in a language of your choice, with the implementation of your choice, and run it on any platform

Comment (brooksmtownsend) [reply wasmcloud-wasi-kv/README.md:14]: Really looking to @ricochet and @autodidaddict here too, this is an important step as people first hit this repository and need to understand components / wasmcloud / wit

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]
Two non-code requests, but I mostly requested changes for cleanliness and language

- Should the Wasm binary files be committed here? Normally we've ignored them but wasn't sure if they were included on purpose
- Can we put this under a top-level folder like `wasi-components` or something more general, rather than this specific folder? Eventually, I'd hope all our examples can be like this

File: wasmcloud-wasi-kv/kv-demo/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        let value = String::from_utf8(value).unwrap();
+        //print("retrieved value: [{value}]");
+
+        // print(format!("setting key [{key}] to value [{value}]").as_str());
+        // keyvalue::set(bucket_handle, key.clone(), value.as_bytes()).expect("failed to set value".into());
+
+        // print("retrieving value for key: [{key}]");
+        // let value = String::from_utf8(keyvalue::get(key).unwrap()).unwrap();
+        // print("retrieved value: [{value}]");
+    }
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/kv-demo/src/lib.rs]: Hey brooks, sorry I think you saw this RIGHT before I pushed up all my changes. 

Have been working out of https://github.com/vados-cosmonic/wamscloud-wasi-kv so PR was very stale!

File: wasmcloud-wasi-kv/kv-demo/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+name = "kv_demo"
+version = "0.1.0"
+edition = "2021"
+publish = false
+
+[lib]
+crate-type = ["cdylib"]
+
+[dependencies]
+wit-bindgen = { git = "https://github.com/bytecodealliance/wit-bindgen" }
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/kv-demo/Cargo.toml:11]: It needs to be [main AFAIK](https://github.com/bytecodealliance/wit-bindgen#guest-rust) -- but I can pin to a specific commit anyway, like [this one (the latest as of the 4d ago)](https://github.com/bytecodealliance/wit-bindgen/commit/7ab30c41b0d537ffc552064156d1c8dca107306e). 

- [ ] pin `main` to commit hash after discussion in sync

This should be coordinated across everyone else's changes though

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs (hunk 1 above)
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:42]: Yeah sure can!

- [ ] Add comments explaining the WIT input/output stream paradigms

File: wasmcloud-wasi-kv/kv-demo/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply wasmcloud-wasi-kv/kv-demo/src/lib.rs]: Ah! Okay feel free to just resolve things that are outdated

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-cc/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "wasmcloud_wasi_kv_cc"
+version = "0.1.0"
+edition = "2021"
+publish = false
+
+[dependencies]
+wit-bindgen = { git = "https://github.com/bytecodealliance/wit-bindgen" }
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-cc/Cargo.toml:8]: Done!

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+publish = false
+
+[lib]
+crate-type = ["cdylib"]
+name = "keyvalue"
+
+[dependencies]
+wasmbus-rpc = "0.11.2" # wasmcloud-interface-keyvalue uses 0.11.2
+wasmcloud-interface-keyvalue = "0.9.1"
+wit-bindgen = { git = "https://github.com/bytecodealliance/wit-bindgen" }
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/Cargo.toml]: Done!

File: wasmcloud-wasi-kv/README.md (hunk 3 above)
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/README.md]: Just refactored! let me know what you think @brooksmtownsend

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs [hunk 8]
```diff
@@ -0,0 +1,52 @@
+
+/// Build a fused component for 
+async fn build_fused_component {
+    
+}
+
+#[tokio::test]
+async fn actor_http_log_rng_component() -> anyhow::Result<()> {
```
Comment (rvolosatovs) [inline wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs:8]: :thinking:

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    let actor = ActorComponent::new(&rt, wasm).expect("failed to read actor component");
+    assert_eq!(actor.claims().subject, key.public_key());
+
+    let mut actor = actor
+        .instantiate()
+        .await
+        .expect("failed to instantiate actor component");
+
+    let response = actor
+        .call("HttpServer.HandleRequest", Some(REQUEST.as_slice()))
```
Comment (rvolosatovs) [inline wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs:34]: pretty sure this is copy-pasted from https://github.com/wasmCloud/wasmCloud/pull/276 and not relevant here

File: wasmcloud-wasi-kv/wit/keyvalue.wit [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+	/// If the key does not exist in the bucket, it returns an error.
+	delete: func(bucket: bucket, key: key) -> result<_, error>
+
+	/// Check if the key exists in the bucket.
+	///
+	/// If the key does not exist in the bucket, it returns an error.
+	exists: func(bucket: bucket, key: key) -> result<bool, error>
+
+	/// Wasmcloud guest
+	guest-call: func(operation: string, payload: option<list<u8>>) -> result<option<list<u8>>, string>
```
Comment (rvolosatovs) [inline wasmcloud-wasi-kv/wit/keyvalue.wit:39]: I'm confused, can't the actor just implement multiple interfaces instead?

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+impl keyvalue::Keyvalue for KvProvider {
+    /// Get a value from the key value store
+    fn get(_bucket: Bucket, key: String) -> Result<IncomingValue, Error> {
+        println!("[debug][kv-provider] performing get...");
+
+        let mut encoder = wasmbus_rpc::cbor::vec_encoder(false);
+        encoder.str(&key)
+            .map_err(|_| ERROR_INTERNAL_PROVIDER_REQUEST_SERIALIZATION_FAILURE)?;
+        let payload = encoder.into_inner();
```
Comment (rvolosatovs) [inline wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:145]: Are you sure this is valid? I though everything that goes over wasmbus uses msgpack

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+    /// Set a value from the key value store
+    fn set(_bucket: Bucket, key: String, outgoing_value: OutgoingValue) -> Result<(), Error> {
+        println!("[debug][kv-provider] performing set...");
+
+        // Read the value from the outgoing value handle
+        //
+        // TODO(QUESTION): It's not strictly necessary to drop an OV after it's used...
+        // It is also a valid use-case to submit the same OV for multiple keys...
+        // Can we trust consumers to requst cleanup of values?
+        let ov_bytes = clone_hashmap_value(&OUTGOING_VALUES, outgoing_value)?.unwrap_or(Vec::new());
```
Comment (rvolosatovs) [inline wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:182]:
```suggestion
        let ov_bytes = clone_hashmap_value(&OUTGOING_VALUES, outgoing_value)?.unwrap_or_default();
```

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs (hunk 8 above)
Comment (t3hmrman) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs:8]: oh woops, yeah this is actually something I was going to use to fill out -- turns out `wasmtime` doesn't build quite so easily here, I had to go back to using your repo

- [ ] remove unfinished tests for keyvalue stuff

Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs:8]: oh woops, yeah this is actually something I was going to use to fill out -- turns out `wasmtime` doesn't build quite so easily here, I had to go back to using your repo

- [ ] remove unfinished tests for keyvalue stuff

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+                // value: unsafe {
+                //     String::from(
+                //         from_utf8_unchecked(&ov_bytes)
+                //             .map_err(|_| ERROR_INTERNAL_PROVIDER_REQUEST_SERIALIZATION_FAILURE)?
+                //     )
+                // },
+                value: String::from(
+                    from_utf8(&ov_bytes)
+                        .map_err(|_| ERROR_INTERNAL_PROVIDER_REQUEST_SERIALIZATION_FAILURE)?,
+                ),
```
Comment (rvolosatovs) [inline wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:203]: I'd extract it into a separate `value` binding and use `.into()` rather than the outer `from` call

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs (hunk 9 above)
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs:34]: Yup, these tests need to be removed all together, `wasmtime` doesn't compile properly with some error about `setjmp.h` being missing.

File: wasmcloud-wasi-kv/wit/keyvalue.wit (hunk 10 above)
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wit/keyvalue.wit:39]: No, https://cosmonic.slack.com/archives/C04S539JRB8/p1678788310748789

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs (hunk 9 above)
Comment (rvolosatovs) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs:34]: :thinking: :thinking: :thinking:

File: wasmcloud-wasi-kv/kv-demo/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+/// Main struct that will implement wasi component imports/exports
+struct Demo;
+
+fn run() -> Result<Option<Vec<u8>>, String> {
+    //print("Running!");
+    println!("Running!");
+    let bucket_name = "bucket";
+    let key = "hello";
+    let value = "wasi";
```
Comment (rvolosatovs) [inline wasmcloud-wasi-kv/kv-demo/src/lib.rs:11]: Do these really help make code easier to maintain and understand?
If you really want these IMO they should be consts instead, i.e. `const BUCKET: &str = "bucket"` etc

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs (hunk 11 above)
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:145]: I'm not sure, but this is the only [encoder that is there](https://docs.rs/wasmbus-rpc/latest/wasmbus_rpc/common/enum.MessageFormat.html?search=encoder). I checked [the other codebases as well](https://github.com/wasmCloud/interfaces/blob/f6c80bb5212dd56dd99d69d579b5b952d894e3c3/testing/rust/src/testing.rs), and it looks to be the right thing.

Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:145]: Hey @autodidaddict can you explain if this would work?

Also, [TIL CBOR was a fork of messagepack](https://github.com/msgpack/msgpack/issues/258)

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+
+        match wasmcloud_host_call(
+            WASMCLOUD_DEFAULT_BINDING,
+            WASMCLOUD_KV_NAMESPACE,
+            "KeyValue.Get",
+            Some(&payload),
+        ) {
+            Ok(Some(resp_bytes)) => {
+                // Parse the get response
+                let resp = wasmbus_rpc::common::deserialize(&resp_bytes)
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:155]: Evidently [CBOR is a fork of msgpack](https://github.com/msgpack/msgpack/issues/258)

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs (hunk 9 above)
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs:34]: I looked and found [this thread](https://github.com/WebAssembly/wasi-sdk/issues/162). I know your branch passes on my machine though, so it doesn't make sense to me but given that I can use your branch to test, I just did that instead.

Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/tests/keyvalue.rs]: test is gone now, so gonna resolve this!

File: wasmcloud-wasi-kv/kv-demo/src/lib.rs (hunk 14 above)
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/kv-demo/src/lib.rs:11]: these are here so they can be reused in the rest of the *demo execution*, I don't think it matters but will change to constants here.

File: wasmcloud-wasi-kv/kv-demo/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,58 @@
+wit_bindgen::generate!("kvdemo" in "../wit");
+
+/// Main struct that will implement wasi component imports/exports
+struct Demo;
+
+fn run() -> Result<Option<Vec<u8>>, String> {
+    //print("Running!");
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/kv-demo/src/lib.rs]: removed this line

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs (hunk 13 above)
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:203]: Oh does that solve the issue with the `String` conversions that Rust does, or is that more of a stylistic change?

File: wasmcloud-wasi-kv/README.md [hunk 17]
```diff
@@ -0,0 +1,120 @@
+# `wasmcloud-wasi-kv`
+
+This repository contains a [WebAssembly System Interface (WASI) key-value contract][wasi-kv] compliant WebAssembly module which can be composed into a "fused" component to provide key value functionality.
+
+This module is special because it implements the KV contract for [wasmCloud][wasmcloud] environments, showcasing the reusability of WASI compliant components across runtimes and platforms.
+
+## ❓ What does that even mean?
+
+- [WebAssembly ("WASM")][wasm] is a compilation target for many languages that creates a platform agnostic binary that can run on any CPU architecture, operating system, or even in the browser
```
Comment (ricochet) [inline wasmcloud-wasi-kv/README.md:9]: Silly nit, but it's "Wasm"

https://webassembly.github.io/spec/core/intro/introduction.html#wasm

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs (hunk 11 above)
Comment (rvolosatovs) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:145]: Sure, but why not just use `wasmbus::rpc::serialize` ?

File: wasmcloud-wasi-kv/README.md (hunk 17 above)
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/README.md:9]: fixed!

File: wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs (hunk 11 above)
Comment (vados-cosmonic) [reply wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs:145]: Ah you're right, will use serialize if it works there!

- [ ] use serialize instead of explicit encoding

File: wasmcloud-wasi-kv/README.md (hunk 3 above)
Comment (brooksmtownsend) [reply wasmcloud-wasi-kv/README.md]: I really like the new bullets. Great stuff @vados-cosmonic
